### PR TITLE
Port cpu_monitor to ROS2

### PR DESCRIPTION
**Pull request by RichardvdK**
I attempted to port the cpu_monitor to ROS2. The module allows users to monitor the CPU usage of their system in real time. It publishes the usage percentage in a diagnostic message.

- I kept the ros arguments warning_percentage and window.
- Added 3 unit tests to cover the code

Let me know what you think! :)

### CODE REVIEW

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+# \author Rein Appeldoorn
+
+import collections
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]:
```suggestion
from collections import deque
```

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+from rclpy.node import Node
+
+
+class CpuTask(DiagnosticTask):
+
+    def __init__(self, warning_percentage=90, window=1):
+        DiagnosticTask.__init__(self, 'CPU Information')
+
+        self._warning_percentage = int(warning_percentage)
+        self._readings = collections.deque(maxlen=window)
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]:
```suggestion
        self._readings = deque(maxlen=window)
```

File: diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            rclpy.shutdown()
+
+    def diagnostics_callback(self, msg):
+        self.message_recieved = True
+        self.assertEqual(len(msg.status), 1)
+
+    def test_ok(self):
+        # In this case is recommended for accuracy that psutil.cpu_percent()
+        # function be called with at least 0.1 seconds between calls.
+        time.sleep(0.1)
```
Comment (ct2034) [inline diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py]: Please move this to a setUp() method (https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp).
Should have afaik the same effect.

File: diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        self.assertEqual(stat.level, DiagnosticStatus.OK)
+        self.assertIn(str('CPU Average'), stat.message)
+
+        # Check for at least 1 CPU Load Average and 1 CPU Load
+        self.assertGreaterEqual(len(stat.values), 2)
+
+    def test_warn(self):
+        # In this case is recommended for accuracy that psutil.cpu_percent()
+        # function be called with at least 0.1 seconds between calls.
+        time.sleep(0.1)
```
Comment (ct2034) [inline diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py]: see line 64

File: diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        self.assertEqual(stat.level, DiagnosticStatus.WARN)
+        self.assertIn(str('At least one CPU exceeds'), stat.message)
+
+        # Check for at least 1 CPU Load Average and 1 CPU Load
+        self.assertGreaterEqual(len(stat.values), 2)
+
+    def test_updater(self):
+        # In this case is recommended for accuracy that psutil.cpu_percent()
+        # function be called with at least 0.1 seconds between calls.
+        time.sleep(0.1)
```
Comment (ct2034) [inline diagnostic_common_diagnostics/test/systemtest/test_cpu_monitor.py]: see line 64

**Review by ct2034 — CHANGES_REQUESTED** [review]
Thanks for porting this. Please see the inline comments

**Comment by RichardvdK** [comment]
> Thanks for porting this. Please see the inline comments

Thanks for reviewing! I updated the files accordingly :)

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py (hunk 2 above)
Comment (ct2034) [reply diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py:58]:
```suggestion
        self._readings = deque(maxlen=window)
```

Comment (ct2034) [reply diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: This is not resolved. Please implement this change

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+        return [avg(cpu_percentages)
+                for cpu_percentages in zip(*self._readings)]
+
+    def run(self, stat):
+        self._readings.append(psutil.cpu_percent(percpu=True))
+        cpu_percentages = self._get_average_reading()
+        cpu_average = sum(cpu_percentages) / len(cpu_percentages)
+
+        stat.add('CPU Load Average', '{:.2f}'.format(cpu_average))
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: Please use f-string

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        warn = False
+        for idx, cpu_percentage in enumerate(cpu_percentages):
+            stat.add('CPU {} Load'.format(idx), '{:.2f}'.format(cpu_percentage))
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: Please use f-string

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            if cpu_percentage > self._warning_percentage:
+                warn = True
+
+        if warn:
+            stat.summary(DiagnosticStatus.WARN,
+                         'At least one CPU exceeds {} percent'.format(self._warning_percentage))
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: Please use f-string

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        else:
+            stat.summary(DiagnosticStatus.OK,
+                         'CPU Average {:.2f} percent'.format(cpu_average))
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: Please use f-string

File: diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+        return stat
+
+
+def main(args=None):
+    rclpy.init(args=args)
+
+    # Create the node
+    hostname = socket.gethostname()
+    node = Node('cpu_monitor_%s' % hostname.replace('-', '_'))
```
Comment (ct2034) [inline diagnostic_common_diagnostics/diagnostic_common_diagnostics/cpu_monitor.py]: Please use f-string

**Review by ct2034 — CHANGES_REQUESTED** [review]
Please fix the failing tests

**Review by ct2034 — CHANGES_REQUESTED** [review]
Please fix 
- SyntaxError in diagnostic_common_diagnostics/cpu_monitor.py:95

**Comment by RichardvdK** [comment]
@ct2034 Excuse moi! Everything should be fixed now :)

File: diagnostic_common_diagnostics/CMakeLists.txt [hunk 11]
```diff
@@ -21,6 +22,10 @@ if(BUILD_TESTING)
         test_ntp_monitor
         test/systemtest/test_ntp_monitor.py
         TIMEOUT 10)
+    ament_add_pytest_test(
+        test_cpu_monitor
+        test/systemtest/test_cpu_monitor.py
```
Comment (ct2034) [inline diagnostic_common_diagnostics/CMakeLists.txt]:
```suggestion
        test_cpu_monitor
        test/systemtest/test_cpu_monitor.py
        TIMEOUT 10)
    ament_add_pytest_test(
        test_ntp_monitor
        test/systemtest/test_ntp_monitor.py
```

Alphabetical ordering

File: diagnostic_common_diagnostics/README.md [hunk 12]
```diff
@@ -8,19 +8,28 @@ Currently only the NTP monitor is ported to ROS2.
...
 
+## cpu_monitor.py
+The `cpu_monitor` module allows users to monitor the CPU usage of their system in real-time.
+It publishes the usage percentage in a diagnostic message.
+
+* Name of the node is "cpu_monitor_" + hostname.
+* Uses the following args:
+  * warning_percentage: If the CPU usage is > warning_percentage, a WARN status will be publised.
+  * window: the maximum length of the used collections.deque for queuing CPU readings.
+
```
Comment (ct2034) [inline diagnostic_common_diagnostics/README.md:18]: Please also move this above ntp_montior. Then add the subsections ### Published Topics and ### Parameters, similarly to how it is done for ntp_monitor

File: diagnostic_common_diagnostics/mainpage.dox [hunk 13]
```diff
@@ -5,6 +5,7 @@
 \b diagnostic_common_diagnostics contains a few common diagnostic nodes
 
 - ntp_monitor publishes diagnostic messages for how well the NTP time sync is working.
+- cpu_monitor publishes diagnostic messages with the CPU usage of the system.
```
Comment (ct2034) [inline diagnostic_common_diagnostics/mainpage.dox]:
```suggestion
- cpu_monitor publishes diagnostic messages with the CPU usage of the system.
- ntp_monitor publishes diagnostic messages for how well the NTP time sync is working.
```

File: diagnostic_common_diagnostics/package.xml [hunk 14]
```diff
@@ -18,21 +18,22 @@
...
-  Once all files are migrated to ROS2, all linters can be 
+  Once all files are migrated to ROS2, all linters can be
   enabled vi ament_lint_common as the files are cleaned up. -->
   <!-- <test_depend>ament_cmake_flake8</test_depend> -->
   <!-- <test_depend>ament_cmake_pep257</test_depend> -->
-  <test_depend>ament_cmake_xmllint</test_depend>
-  <test_depend>ament_cmake_lint_cmake</test_depend>
-  
-  <test_depend>ament_cmake_pytest</test_depend>
+
```
Comment (ct2034) [inline diagnostic_common_diagnostics/package.xml]: Please reduce the number of changes here to what is necessary

**Review by ct2034 — CHANGES_REQUESTED** [review]
Sorry, I wanted to make these changes myself but could not push to your branch

**Comment by RichardvdK** [comment]
@ct2034 did you find time already to re-review this PR? :)

**Review by ct2034 — APPROVED** [review]

**Comment by ct2034** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|ros2-jazzy|[](https://github.com/ros/diagnostics/pull/385)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**